### ISSUE TITLE
multiple ChartProviders may render selector memoization useless

### ISSUE THREAD

**Issue by seansfkelley** [palantir/react-layered-chart#18]
Default memoization has a cache size of 1: https://github.com/reactjs/reselect#defaultmemoizefunc-equalitycheck--defaultequalitycheck

If there are multiple ChartProviders sharing the same set of selectors, I think this means that they will clobber each others' cached results any time any one of them changes.

**Comment by seansfkelley** [comment]
[`mapStateToProps` can be a factory function too](https://github.com/reactjs/react-redux/blob/master/docs/api.md#arguments), so we can use per-instance memoization in some fashion.

**Comment by zakmagnus** [comment]
I think I've observed this issue. I added some console activity to the selectors so I could easily see when recomputation was happening, then ran the example page and panned around on one of the charts. I saw a bunch more recomputation than I would expect, and I suspect it's exactly because of this issue: the charts are repeatedly invalidating the one cache they all share.

**Comment by seansfkelley** [comment]
Should we then switch to the factory-function version of `mapStateToProps`?

**Comment by zakmagnus** [comment]
Yes, or something along those lines. I suspect that's only part of the problem. Mainly, I think the issue is that each selector is just a `const` object, and that object is also what does the caching, so I think every component which uses a selector will use the same cache. So, I think there needs to be a way to instantiate the selectors for a specific chart. Once that is possible, there could be some sort of `mapStateToPropsMaker` which creates the selectors and reuses them as appropriate.

If this is correct, it has wider implications, since right now arbitrary children of `ChartProvider` use selectors to grab data; but actually they need to use the "instantiated selectors" for the chart they're under, not some global ones.

I'm hardly any expert on typescript scoping rules and such, so I may be mistaken.

**Comment by zakmagnus** [comment]
An alternative is to put a "selector object" into the store itself. Then, global selectors can continue to exist, but all they do is delegate to the object inside the store. The object has to be created and injected into the state in `storeFactory`. Something like this:

```
function _createStore(...) {
  ...
  const initialState = { selectors: createSelectors() };
  return createStore(...)(initialState);
}

function createSelectors() {
  return {
    selectXDomain: createSelector(...)
    ...
  };
}

...

const selectState = (state) => state;
const selectSelectors = (state) => state.selectors;
export const selectXDomain = createSelector(
  selectState,
  selectSelectors,
  (state, selectors) => selectors.selectXDomain(state)
);
```

It seems strange to put something like this into redux state, but it is convenient.
